database/detector.go

```diff
@@ -0,0 +1,144 @@
+	FeatureDetectorType DetectorType = "DETECTOR_TYPE_FEATURE"
+)
+
+// DetectorTypes contains all detector types.
+var (
+	DetectorTypes = []DetectorType{
+		NamespaceDetectorType,
+		FeatureDetectorType,
+	}
+	// ErrFailedToParseDetectorType is the error returned when a detector type could
+	// not be parsed from a string.
+	ErrFailedToParseDetectorType = errors.New("failed to parse DetectorType from input")
+	// ErrInvalidDetector is the error returned when a detector from database has
+	// invalid name or version or type.
+	ErrInvalidDetector = errors.New("the detector has invalid metadata")
+)
+
+// DetectorType is the type of a detector.
+type DetectorType string
+
+// Value implements the database/sql/driver.Valuer interface.
+func (s DetectorType) Value() (driver.Value, error) {
+	return string(s), nil
+}
+
+// Scan implements the database/sql.Scanner interface.
+func (s *DetectorType) Scan(value interface{}) error {
+	val, ok := value.([]byte)
+	if !ok {
+		return errors.New("could not scan a Severity from a non-string input")
+	}
+
+	var err error
+	*s, err = NewDetectorType(string(val))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// NewDetectorType attempts to parse a string into a standard DetectorType
+// value.
+func NewDetectorType(s string) (DetectorType, error) {
+	for _, ss := range DetectorTypes {
+		if strings.EqualFold(s, string(ss)) {
+			return ss, nil
+		}
+	}
+
+	return "", ErrFailedToParseSeverity
+}
+
+// Valid checks if a detector type is defined.
+func (s DetectorType) Valid() bool {
+	for _, t := range DetectorTypes {
+		if s == t {
+			return true
+		}
+	}
+
+	return false
+}
+
+// Detector is an versioned Clair extension.
+type Detector struct {
+	// Name of an extension should be non-empty and uniquely identifies the
+	// extension.
+	Name string
+	// Version of an extension should be non-empty.
+	Version string
+	// DType is the type of the extension and should be one of the types in
+	// DetectorTypes.
+	DType DetectorType
+}
+
+// Valid checks if all fields in the detector satisfies the spec.
+func (d Detector) Valid() bool {
+	if d.Name == "" || d.Version == "" || !d.DType.Valid() {
+		return false
+	}
+
+	return true
+}
+
+// String returns a unique string representation of the detector.
+func (d Detector) String() string {
+	return fmt.Sprintf("%sDetector/%s/%s", d.DType, d.Name, d.Version)
+}
+
+// NewNamespaceDetector returns a new namespace detector.
+func NewNamespaceDetector(name, version string) Detector {
+	return Detector{
+		Name:    name,
+		Version: version,
+		DType:   NamespaceDetectorType,
+	}
+}
+
+// NewFeatureDetector returns a new feature detector.
+func NewFeatureDetector(name, version string) Detector {
+	return Detector{
+		Name:    name,
+		Version: version,
+		DType:   FeatureDetectorType,
+	}
+}
+
+// SerializeDetectors returns the string representation of given detectors.
+func SerializeDetectors(detectors []Detector) []string {
```

Comment:
ok, then it will require the type conversion from []database.Detector to []fmt.Stringer, which sucks

Reply:
Yeah. This is super unfortunate... It's because interfaces different memory layouts and creating a list of them is O(N), so you have to explicitly make the copy yourself rather than having Go do it for you when you pass it into the function.

Reply:
It makes me think if this kind of design is wrong after all. I think we are thinking about functions using *generics* but interfaces are not generics.

Reply:
I think I should keep the original design for the serialization.

Reply:
:thumbsup: